From 87f9a90784c10aa4dd07869380b85596ff20976d Mon Sep 17 00:00:00 2001
From: William Bright <William.Bright@imd-tec.com>
Date: Wed, 15 Apr 2026 11:57:54 +0100
Subject: [PATCH] arm64: dts: qcom: qcs8550-imdt-sbc: Add PCIe switch Key B
 overlay

Add a device tree overlay to select M.2 Key B PCIe routing instead of
the default GBE PHY routing. The overlay overrides the pcie_switch_sel_gpio
hog to invert the GPIO polarity, driving the physical pin high.

Add a label to the PCIe switch GPIO hog node so the overlay can
reference it.

Signed-off-by: William Bright <william.bright@imd-tec.com>
---
 arch/arm64/boot/dts/qcom/Makefile             |  1 +
 .../dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso  | 20 +++++++++++++++++++
 arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc.dts | 10 ++++++++++
 3 files changed, 31 insertions(+)
 create mode 100644 arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso

diff --git a/arch/arm64/boot/dts/qcom/Makefile b/arch/arm64/boot/dts/qcom/Makefile
index 79fd05a32e00..a9a57392b4c9 100644
--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -221,6 +221,7 @@ dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-aim300-aiot.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-display.dtbo
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-ar1335-csi0.dtbo
+dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-pcie-keyb.dtbo
 
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs9100-ride.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs9100-ride-r3.dtb
diff --git a/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso
new file mode 100644
index 000000000000..e07f44bf6bdb
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2026 IMD Technologies Ltd
+ *
+ * There are two options for PCIe routing on the IMDT QCS8550 SBC.
+ * By default, the PCIe switch is configured to route to the GBE PHY, but
+ * this can be changed to route to the M.2 slot Key B by setting 
+ * pcie_switch_sel_gpio high
+ */
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&pcie_switch_sel_gpio {
+		/* Change the PCIe switch so that key B routing is selected by
+		 * inverting the GPIO polarity.
+		 */
+		gpios = <16 GPIO_ACTIVE_LOW>;
+};
diff --git a/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc.dts b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc.dts
index 73a13c4f0e65..bd3ef3567874 100644
--- a/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc.dts
+++ b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc.dts
@@ -770,6 +770,16 @@ m2e_sdio_resetn_default: m2e-sdio-resetn-default-state {
 		output-high;
 	};
 
+	/* Force PCIE switch to select GBE PHY */
+	pcie_switch_sel_gpio: pcie-switch-sel-gpio {
+		gpio-hog;
+		gpios = <16 GPIO_ACTIVE_HIGH>;
+
+		drive-strength = <2>;
+		bias-disable;
+		output-high;
+	};
+
 	/* Force the on-board PCIe switch to select the GbE upstream port. */
 	pcie_switch_sel_default: pcie-switch-sel-default-state {
 		pins = "gpio16";
-- 
2.43.0

